Automattic/newspack-popups
# Polyfill amp-analytics script

### All Submissions:






### Changes proposed in this Pull Request:



Polyfill of `amp-analytics` AMP script (just the bits used by the plugin), part of #193.

A part that needs some explanation here is the linker parameter. Here's [a short overview](https://amp.dev/documentation/components/amp-analytics/#linkers), and here the [documentation about linker param receiving](https://github.com/ampproject/amphtml/blob/master/extensions/amp-analytics/linker-id-receiving.md) – the part that's actually polyfilled. Linker param _setting_ is not handled here, because it would be set only by a page served from AMP cache, which would be… an AMP page (not a version with polyfill).

Here's an example of a linker param:
```
?ref_newspack_cid=1*10zr7ce*cid*T2ptZDdJNU9EYndELWV0TlM1N0FwSGVwNHE3S240VkVVU0o3YlNsaFc2YVVPRXJhSWFhaVlsOTgtQXNsc21Eeg
```

The `T2pt…` part is base64-encoded client ID (of value `Ojmd7I5ODbwD-etNS57ApHep4q7Kn4VEUSJ7bSlhW6aUOEraIaaiYl98-AslsmDz`). You can this set the client ID to a specified value (in JS, e.g. `btoa('hello-world')` gives you a base64 encoded `hello-world`).

Note that the checksum part of the linker param is disregarded in this polyfill. I think the security implications of client ID spoofing in this use case are negligible.

### How to test the changes in this Pull Request:

1. Linker parameter
    1. Append a linker param to a post URL in non-AMP mode, visit the site in a fresh session (with no `newspack-cid` cookie set)
    2. Observe in the devtools that the `newspack-cid` cookie is set accordingly
    3. Observe that the linker param is stripped from the URL afterwards
    4. Confirm that this works in the same way on an AMP site – you can use [this AMP-cached site](https://campaigns--testing-newspackstaging-com.cdn.ampproject.org/c/s/campaigns-testing.newspackstaging.com/) to "generate" linker params with valid checksums
2. Reporting views to API 
    1. Create a campaign with "once" frequency
    2. Visit the site in a fresh session in non-AMP mode, observe the campaign is displayed (make sure you "see" it – that means at least 500ms with 90% of the campaign visible in the viewport)
    3. Reload the site or navigate to another post – observe the campaign is not displayed

### Other information:
